Incorporate changes to `main.tex` and `main.bib` from Karl Berry (#23)
Witiko/fast-regression-testing: main.tex

\section{Introduction}
Small \TeX{} packages, typically developed in a single iteration rather than through ongoing updates, can depend on user feedback to maintain the code. However, this approach has its limitations. Larger projects, especially those that are continuously developed, require a more robust solution. Automated regression tests are crucial in these cases. They ensure that any changes, either in the code itself or its external dependencies, do not alter the expected behavior of the code.

> The Markdown package for \TeX{} features a~set of regression tests. These tests, designed to be completed in just a few minutes, provide immediate feedback and are automatically conducted on any updates submitted to the package's GitHub repository.

Author: The "also" refers to the previous paragraph. Larger projects require automated regression tests. The Markdown package for TeX is a larger project. Therefore, it also features regression tests.

```suggestion
The Markdown package for \TeX{} also features a~set of regression tests. These tests, designed to be completed in just a few minutes, provide immediate feedback and are automatically conducted on any updates submitted to the package's GitHub repository.
```

A \emph{test file} consists of a) \TeX{} code that configures the Markdown package, b) markdown text, and c)~the expected output in the \texttt{.log} file.

> As an example, \file{strike-through.test} tests the strike-through syntax extension:

Author: I assume that the change was made to improve the line breaking after the removal of inline icons. However, since we wish to keep the inline icons, this change seems unnecessary and inconsistent with other similar examples throughout the text.

```suggestion
An example test file \file{strike-through.test} tests the strike-through syntax extension:
```

\noindent
\addtocounter{footnote}{1}%
\footnotetext{%

>   The Markdown package uses Lua to parse markdown text. Whereas Lua\TeX{} can execute Lua code directly, other \TeX{} engines must use the operating system shell to execute Lua code. Since accessing the shell is a security risk, users must express their consent by writing \texttt{-{}-shell-escape}.%

Author: I prefer "shell of the operating system" rather than "operating system shell". For the less technical reader, it separates the two concepts (shell and operating system), which makes the text easier to digest. The line breaking does not seem to be improved either way.

```suggestion
  The Markdown package uses Lua to parse markdown text. Whereas Lua\TeX{} can execute Lua code directly, other \TeX{} engines must use the shell of the operating system to execute Lua code. Since accessing the shell is a security risk, users must express their consent by writing \texttt{-{}-shell-escape}.%
```

\section{Implementation}
\label{sec:implementation}

> Before Markdown 3.0.0, our framework was implemented by a Bash script \file{test.sh}, shown in Listing~\ref{lst:test.sh}.

Author: This change seems unnecessary and produces an extra short line that looks unseemly.

```suggestion
Before Markdown 3.0.0, our framework was implemented by a Bash script \file{test.sh}, see Listing~\ref{lst:test.sh}.
```

Figure~\ref{fig:results} shows the results of our experiments. In this section, we discuss the results and how they relate to the three research questions outlined in the previous section.

\subsection{Multiprocessing}

> With batch size 1, the testing speed scales almost linearly with the number of \acro{CPU}s, as we would expect: with 1~\acro{CPU}, the median testing time is 27~hours and 2~minutes, it is only 1~hour and 6~minutes with 32~\acro{CPU}s (about 24-fold speed-up).\footnote{%

Author: I assume that you left out "Whereas" here. Without it, the "it is only ..." part of the last clause seems misplaced.

```suggestion
With batch size 1, the testing speed scales almost linearly with the number of \acro{CPU}s, as we would expect: Whereas with 1~\acro{CPU}, the median testing time is 27~hours and 2~minutes, it is only 1~hour and 6~minutes with 32~\acro{CPU}s (about 24-fold speed-up).\footnote{%
```

The reason that we did not achieve the theoretical 32-fold speedup is likely tasks from other users on our server.%
}

\subsection{Batching of test files}

> With 1~\acro{CPU}, the testing speed also scales almost linearly with the batch size, up to a point. Whereas with batch size 1, the median testing time is 27~hours and 2~minutes, it is 4~hours and 30~minutes with batch size 8 (about 6-fold speed-up), and 1~hour and 20~minutes with batch size 512 (about 21-fold speed-up). This indicates that initialization dominates the testing time.

Author: The removal of "only" is inconsistent with the previous section and also next section.

```suggestion
With 1~\acro{CPU}, the testing speed also scales almost linearly with the batch size, up to a point. Whereas with batch size 1, the median testing time is 27~hours and 2~minutes, it is only 4~hours and 30~minutes with batch size 8 (about 6-fold speed-up), and 1~hour and 20~minutes with batch size 512 (about 21-fold speed-up). This indicates that initialization dominates the testing time.
```

When the number of \acro{CPU}s multiplied by the batch size exceeds the number of test files (783), we cannot use all \acro{CPU}s and the testing speed decreases. Whereas with 32~\acro{CPU}s and batch size 16, the median testing time is only 11~minutes, it is 1~hour and 24~minutes with the same number of \acro{CPU}s and batch size 1024 (about 8-fold slow-down). Our framework prevents this effect by limiting the batch size.

\section{Related work}
\label{sec:related-work}

Besides our framework, there exist other frameworks for regression testing of \TeX{} packages. Furthermore, the computational techniques in our framework are often adapted from previous work in other fields.

In this section, we discuss the regression testing framework of the l3build package management system and we compare it with our framework. Furthermore, we discuss the origin of the batching of test files and batch splitting.

\subsection{The l3build package}

> The l3build package~\cite{mittelbach2014l3build, latex2023l3build, wright2022l3build, wright2015automating} provides a comprehensive system for \TeX{} package management that also includes a regression testing framework.

Author: I list the citations in the order of publication.
```suggestion
The l3build package~\cite{mittelbach2014l3build, wright2015automating, wright2022l3build, latex2023l3build} provides a comprehensive system for \TeX{} package management that also includes a regression testing framework.
```

Both our framework and l3build support the updating of test files~\cite[Section~2.7]{latex2023l3build}. This allows developers to automatically generate parts of test files when they make fundamental changes to the code or when they develop complex new features, as discussed in Section~\ref{sec:updating-test-files}.

\subsection{Batching and batch splitting}
The techniques of batching and batch splitting were perhaps first used with \TeX{} in the \ARQMath\ competitions in large-scale indexing of math formulae.

> In the first \ARQMath\ competition, the \acro{MIRMU} team used \LaTeX\acro{ML} to convert \TeX{} formulae to \acro{XML}~\cite[Section~2.2]{novotny2020three}. Due to speed issues when processing each formula separately, \acro{MIRMU} processed them in batches. However, a single error would cause the loss of an entire batch. Therefore, \acro{MIRMU} used batch splitting to recover correct formulae after an error~\cite{novotny2020arqmath}.

Author: Since this is the first occurence of LaTeXML in the text, let's help the reader by qualifying LaTeXML as "a tool". Ommitting it does not seem to improve the line breaks.
```suggestion
In the first \ARQMath\ competition, the \acro{MIRMU} team used the \LaTeX\acro{ML} tool to convert \TeX{} formulae to \acro{XML}~\cite[Section~2.2]{novotny2020three}. Due to speed issues when processing each formula separately, \acro{MIRMU} processed them in batches. However, a single error would cause the loss of an entire batch. Therefore, \acro{MIRMU} used batch splitting to recover correct formulae after an error~\cite{novotny2020arqmath}.
```
